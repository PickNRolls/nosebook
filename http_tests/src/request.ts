import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
export const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;
  private auth: boolean;

  public constructor(host: string, auth?: boolean) {
    this.host = host;
    this.agent = supertest(host);
    this.auth = auth ?? true;
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(opts?: {
    prefixUrl?: string;
    auth?: boolean;
  }): TestAgentWrapper {
    let url = this.host;
    if (opts?.prefixUrl) {
      url += opts.prefixUrl;
    }
    return new TestAgentWrapper(url, opts?.auth ?? this.auth);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    if (this.auth) {
      test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');
    }

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


