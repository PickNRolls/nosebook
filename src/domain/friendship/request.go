package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
