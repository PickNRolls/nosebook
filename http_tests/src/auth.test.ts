import supertest from 'supertest';
import request, { SESSION_HEADER_KEY } from './request';

describe('auth', () => {
  let auth = request.extend({
    auth: false,
  });

  test('GET /whoami', async () => {
    let response = await auth
      .get('/whoami')
      .expect(403)

    expect(response.body).toMatchSnapshot();
  });

  test('POST /register', async () => {
    let response = await auth
      .post('/register')
      .send({
        firstName: 'test',
        lastName: 'test',
        nick: 'test_nick2',
        password: '123123123',
      });

    expect(response.body).toMatchObject({
      ok: true,
      data: {
        user: {
          id: expect.any(String),
          firstName: 'test',
          lastName: 'test',
          nick: 'test_nick2',
          passhash: expect.any(String),
          createdAt: expect.any(String),
        },
        session: {
          sessionId: expect.any(String),
          userId: expect.any(String),
          createdAt: expect.any(String),
          expiresAt: expect.any(String),
        },
      }
    });
  });

  test('POST /login', async () => {
    let response = await auth
      .post('/login')
      .send({
        nick: 'test_tester',
        password: '123123123',
      });

    expect(response.body).toMatchObject({
      ok: true,
      data: {
        user: {
          id: expect.any(String),
          firstName: 'Test',
          lastName: 'Tester',
          nick: 'test_tester',
          passhash: expect.any(String),
          createdAt: expect.any(String),
          lastActivityAt: expect.any(String),
        },
        session: {
          sessionId: expect.any(String),
          userId: expect.any(String),
          createdAt: expect.any(String),
          expiresAt: expect.any(String),
        },
      }
    });
  });

  test('POST /logout', async () => {
    let response = await auth
      .post('/logout')
      .set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e55');

    expect(response.body).toMatchObject({
      ok: true,
      data: {
        sessionId: 'bb23af03-be50-4bce-b729-b259b2e02e55',
        userId: expect.any(String),
        createdAt: expect.any(String),
        expiresAt: expect.any(String),
      }
    });
  });
});

