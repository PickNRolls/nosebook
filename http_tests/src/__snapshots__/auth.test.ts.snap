// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth GET /whoami 1`] = `
{
  "errors": [
    {
      "message": "You are not authenticated",
      "type": "Not Authenticated",
    },
  ],
  "ok": false,
}
`;
